src/Microsoft.Net.Http.Headers/HttpRuleParser.cs

```diff
@@ -0,0 +1,394 @@
+            TokenChars[(byte)'['] = false;
+            TokenChars[(byte)']'] = false;
+            TokenChars[(byte)'?'] = false;
+            TokenChars[(byte)'='] = false;
+            TokenChars[(byte)'{'] = false;
+            TokenChars[(byte)'}'] = false;
+        }
+
+        internal static bool IsTokenChar(char character)
+        {
+            // Must be between 'space' (32) and 'DEL' (127)
+            if (character > 127)
+            {
+                return false;
+            }
+
+            return TokenChars[character];
+        }
+
+        [Pure]
+        internal static int GetTokenLength(string input, int startIndex)
+        {
+            Contract.Requires(input != null);
+            Contract.Ensures((Contract.Result<int>() >= 0) && (Contract.Result<int>() <= (input.Length - startIndex)));
+
+            if (startIndex >= input.Length)
+            {
+                return 0;
+            }
+
+            var current = startIndex;
+
+            while (current < input.Length)
+            {
+                if (!IsTokenChar(input[current]))
+                {
+                    return current - startIndex;
+                }
+                current++;
+            }
+            return input.Length - startIndex;
+        }
+
+        internal static int GetWhitespaceLength(string input, int startIndex)
+        {
+            Contract.Requires(input != null);
+            Contract.Ensures((Contract.Result<int>() >= 0) && (Contract.Result<int>() <= (input.Length - startIndex)));
+
+            if (startIndex >= input.Length)
+            {
+                return 0;
+            }
+
+            var current = startIndex;
+
+            char c;
+            while (current < input.Length)
+            {
+                c = input[current];
+
+                if ((c == ' ') || (c == '\t'))
+                {
+                    current++;
+                    continue;
+                }
+
+                if (c == '\r')
+                {
+                    // If we have a #13 char, it must be followed by #10 and then at least one SP or HT.
+                    if ((current + 2 < input.Length) && (input[current + 1] == '\n'))
+                    {
+                        char spaceOrTab = input[current + 2];
+                        if ((spaceOrTab == ' ') || (spaceOrTab == '\t'))
+                        {
+                            current += 3;
+                            continue;
+                        }
+                    }
+                }
+
+                return current - startIndex;
+            }
+
+            // All characters between startIndex and the end of the string are LWS characters.
+            return input.Length - startIndex;
+        }
+
+        internal static bool ContainsInvalidNewLine(string value)
+        {
+            return ContainsInvalidNewLine(value, 0);
+        }
+
+        internal static bool ContainsInvalidNewLine(string value, int startIndex)
```

Comment:
Do we need to check for \n without \r?  I realize it's allowable per spec, but is there any legitimate scenario for accepting such a thing or is it really an attempt to confuse the server into doing something strange?  I'm mainly concerned about UNIX scenarios where the environment newline character is \n instead of the \r\n combo.